Support for Android API Level < 24?

The library (version 0.4.2) seems to leverage java.net.StandardProtocolFamily and java.net.StandardSocketOptions which are only available at API level 24 and greater. Is there a version of JeroMQ that operates on a lower API level then this (with Curve support)?

It doesn't make it to that call, it stops at setReuseAddress():

`Caused by: java.lang.ClassNotFoundException: Didn't find class "java.nio.channels.NetworkChannel" on path: DexPathList[[zip file "/data/app/com.futureskies.jeromqtest-1/base.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_dependencies_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_0_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_1_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_2_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_3_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_4_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_5_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_6_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_7_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_8_apk.apk", zip file "/data/app/com.futureskies.jeromqtest-1/split_lib_slice_9_apk.apk"],nativeLibraryDirectories=[/vendor/lib64, /system/lib64]]
                                                                              at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                                                              at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
                                                                              at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
                                                                              at zmq.io.net.tcp.TcpListener.setAddress(TcpListener.java:186) 
                                                                              at zmq.SocketBase.bind(SocketBase.java:363) 
                                                                              at org.zeromq.ZMQ$Socket.bind(ZMQ.java:2333) `

--- a/src/main/java/zmq/io/Metadata.java
+++ b/src/main/java/zmq/io/Metadata.java
@@ -1,6 +1,5 @@
 package zmq.io;
 
-import java.util.Objects;
 import java.util.Properties;
 
 public class Metadata
@@ -41,17 +40,23 @@ public class Metadata
     @Override
     public int hashCode()
     {
-        return Objects.hashCode(dictionary);
+        return dictionary.hashCode();
     }
 
     @Override
-    public boolean equals(Object obj)
+    public boolean equals(Object other)
     {
-        if (obj instanceof Metadata) {
-            Metadata other = (Metadata) obj;
-            return dictionary.equals(other.dictionary);
+        if (this == other) {
+            return true;
         }
-        return false;
+        if (other == null) {
+            return false;
+        }
+        if (!(other instanceof Metadata)) {
+            return false;
+        }
+        Metadata that = (Metadata) other;
+        return this.dictionary.equals(that.dictionary);
     }
 
     public final void set(Metadata zapProperties)

--- a/src/main/java/zmq/io/net/tcp/TcpListener.java
+++ b/src/main/java/zmq/io/net/tcp/TcpListener.java
@@ -187,7 +187,8 @@ public class TcpListener extends Own implements IPollEvents
             }
 
             //  Bind the socket to the network interface and port.
-            fd.bind(address.address(), options.backlog);
+            // NB: fd.socket().bind(...) for Android environments
+            fd.socket().bind(address.address(), options.backlog);
         }
         catch (IOException e) {
             close();

--- a/src/main/java/zmq/poll/Poller.java
+++ b/src/main/java/zmq/poll/Poller.java
@@ -9,7 +9,6 @@ import java.nio.channels.SelectionKey;
 import java.nio.channels.Selector;
 import java.util.HashSet;
 import java.util.Iterator;
-import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -31,6 +30,8 @@ public final class Poller extends PollerBase implements Runnable
 
         public Handle(SelectableChannel fd, IPollEvents handler)
         {
+            assert (fd != null);
+            assert (handler != null);
             this.fd = fd;
             this.handler = handler;
         }
@@ -38,23 +39,27 @@ public final class Poller extends PollerBase implements Runnable
         @Override
         public int hashCode()
         {
-            return Objects.hash(fd, handler);
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + fd.hashCode();
+            result = prime * result + handler.hashCode();
+            return result;
         }
 
         @Override
-        public boolean equals(Object obj)
+        public boolean equals(Object other)
         {
-            if (this == obj) {
+            if (this == other) {
                 return true;
             }
-            if (obj == null) {
+            if (other == null) {
                 return false;
             }
-            if (!(obj instanceof Handle)) {
+            if (!(other instanceof Handle)) {
                 return false;
             }
-            Handle other = (Handle) obj;
-            return Objects.equals(fd, other.fd) && Objects.equals(handler, other.handler);
+            Handle that = (Handle) other;
+            return this.fd.equals(that.fd) && this.handler.equals(that.handler);
         }
 
         @Override

--- a/src/main/java/zmq/poll/PollerBase.java
+++ b/src/main/java/zmq/poll/PollerBase.java
@@ -1,7 +1,6 @@
 package zmq.poll;
 
 import java.util.Map.Entry;
-import java.util.Objects;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import zmq.util.Clock;
@@ -18,6 +17,7 @@ abstract class PollerBase implements Runnable
 
         public TimerInfo(IPollEvents sink, int id)
         {
+            assert (sink != null);
             this.sink = sink;
             this.id = id;
         }
@@ -25,17 +25,27 @@ abstract class PollerBase implements Runnable
         @Override
         public int hashCode()
         {
-            return Objects.hash(id, sink);
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + id;
+            result = prime * result + sink.hashCode();
+            return result;
         }
 
         @Override
-        public boolean equals(Object obj)
+        public boolean equals(Object other)
         {
-            if (obj instanceof TimerInfo) {
-                TimerInfo other = (TimerInfo) obj;
-                return this.id == other.id && Objects.equals(this.sink, other.sink);
+            if (this == other) {
+                return true;
             }
-            return false;
+            if (other == null) {
+                return false;
+            }
+            if (!(other instanceof TimerInfo)) {
+                return false;
+            }
+            TimerInfo that = (TimerInfo) other;
+            return this.id == that.id && this.sink.equals(that.sink);
         }
 
         @Override

--- a/src/main/java/zmq/util/ValueReference.java
+++ b/src/main/java/zmq/util/ValueReference.java
@@ -1,7 +1,5 @@
 package zmq.util;
 
-import java.util.Objects;
-
 public class ValueReference<V>
 {
     private V value;
@@ -28,6 +26,6 @@ public class ValueReference<V>
     @Override
     public String toString()
     {
-        return Objects.toString(value);
+        return value == null ? "null" : value.toString();
     }
 }

